multisig-wallet/src/main.sw

```diff
@@ -0,0 +1,160 @@
+contract;
+
+use std::{address::Address, assert::assert, chain::auth::{AuthError, Sender, msg_sender}, contract_id::ContractId, hash::{HashMethod, hash_pair, hash_u64}, panic::panic, result::*, storage::{get, store}};
+
+abi MultiSignatureWallet {
+    fn constructor(owner1: Address, owner2: Address, threshold: u64) -> bool;
+    fn submit(destination: ContractId, value: u64, data: b256) -> bool;
+    fn approve(nonce: u64) -> bool;
+    fn revokeApproval(nonce: u64) -> bool;
+    fn execute(nonce: u64) -> bool;
+    fn isOwner(owner: Address) -> bool;
+    fn getTransaction(nonce: u64) -> (b256, u64, b256, bool, u64);
+}
+
+struct SentinelTransaction {
+    to: b256,
+    value: b256,
+    data: b256,
+    executed: b256,
+    approvals: b256,
+}
+
+// TODO: add logging events
+
+storage {
+    state: u64,
+    nonce: u64,
+    threshold: u64,
+    sentinel: SentinelTransaction,
+}
+
+impl MultiSignatureWallet for Contract {
+    fn constructor(owner1: Address, owner2: Address, threshold: u64) -> bool {
+        assert(storage.state == 0);
+
+        // TODO: when vectors are implemented change owners to be a Vec<Address>
+        store(owner1.value, true);
+        store(owner2.value, true);
+
+        storage.sentinel = SentinelTransaction {
+            to: hash_u64(1, HashMethod::Sha256), value: hash_u64(2, HashMethod::Sha256), data: hash_u64(3, HashMethod::Sha256), executed: hash_u64(4, HashMethod::Sha256), approvals: hash_u64(5, HashMethod::Sha256)
+        };
+
+        storage.threshold = threshold;
+        storage.state = 1;
+        true
+    }
+
+    fn submit(destination: ContractId, value: u64, data: b256) -> bool {
+        assert(storage.state == 1);
+
+        let sender: Result<Sender, AuthError> = msg_sender();
+        if let Sender::Address(address) = sender.unwrap() {
+            assert(get(address.value));
+
+            storage.nonce = storage.nonce + 1;
+
+            let nonce_hash = hash_u64(storage.nonce, HashMethod::Sha256);
+
+            store(hash_pair(nonce_hash, storage.sentinel.to, HashMethod::Sha256), destination.value);
+            store(hash_pair(nonce_hash, storage.sentinel.value, HashMethod::Sha256), value);
+            store(hash_pair(nonce_hash, storage.sentinel.data, HashMethod::Sha256), data);
+        } else {
+            panic(0);
+        };
+
+        true
+    }
+
+    fn approve(nonce: u64) -> bool {
```

Comment:
Approvals should be done over transaction hashes (which includes a nonce) rather than a bare nonce. The latter is vulnerable to re-orgs changing what the nonce represents.